src/storage/src/storage/signed_url.rs

```diff
@@ -0,0 +1,184 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::SigningError;
+use auth::signer::Signer;
+use std::collections::BTreeMap;
+use tokio::time::Duration;
+
+/// A builder for creating signed URLs.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct SignedUrlBuilder {
+    scope: SigningScope,
+    method: http::Method,
+    expiration: std::time::Duration,
+    headers: BTreeMap<String, String>,
+    query_parameters: BTreeMap<String, String>,
+    endpoint: Option<String>,
+    universe_domain: String,
+    client_email: Option<String>,
+    url_style: UrlStyle,
+}
+
+#[derive(Debug, Clone, Default)]
+#[allow(dead_code)]
+pub enum UrlStyle {
+    #[default]
+    PathStyle,
+    BucketBoundHostname,
+    VirtualHostedStyle,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+enum SigningScope {
+    Bucket(String),
+    Object(String, String),
+}
+
+impl SignedUrlBuilder {
+    fn new(scope: SigningScope) -> Self {
+        Self {
+            scope,
+            method: http::Method::GET,
+            expiration: std::time::Duration::from_secs(7 * 24 * 60 * 60), // 7 days
+            headers: BTreeMap::new(),
+            query_parameters: BTreeMap::new(),
+            endpoint: None,
+            universe_domain: "googleapis.com".to_string(),
+            client_email: None,
+            url_style: UrlStyle::PathStyle,
+        }
+    }
+
+    pub fn for_object<B, O>(bucket: B, object: O) -> Self
+    where
+        B: Into<String>,
+        O: Into<String>,
+    {
+        Self::new(SigningScope::Object(bucket.into(), object.into()))
+    }
+
+    pub fn for_bucket<B>(bucket: B) -> Self
+    where
+        B: Into<String>,
+    {
+        Self::new(SigningScope::Bucket(bucket.into()))
+    }
+
+    /// Sets the HTTP method for the signed URL. Default is "GET".
+    pub fn with_method(mut self, method: http::Method) -> Self {
+        self.method = method;
+        self
+    }
+
+    /// Sets the expiration time for the signed URL. Default is 7 days.
+    pub fn with_expiration(mut self, expiration: Duration) -> Self {
+        self.expiration = expiration;
+        self
+    }
+
+    /// Sets the URL style for the signed URL. Default is `UrlStyle::PathStyle`.
+    pub fn with_url_style(mut self, url_style: UrlStyle) -> Self {
+        self.url_style = url_style;
+        self
+    }
+
+    /// Adds a header to the signed URL.
+    /// Note: These headers must be present in the request when using the signed URL.
```

Comment:
Separate the title from the notes:


```suggestion
    /// Adds a header to the signed URL.
    ///
    /// Note: These headers must be present in the request when using the signed URL.
```